### ISSUE TITLE
A debugging check in this application has Failed

### ISSUE THREAD

**Issue by LezheGao** [BOINC/boinc#4806]
![1](https://user-images.githubusercontent.com/76685272/174992738-771c0a37-9754-4eb7-9b7d-07758227b010.jpg)

**Comment by AenBleidd** [comment]
Closed as a duplicate of #4783

**Comment by LezheGao** [comment]
> 如# 4783[的副本关闭](https://github.com/BOINC/boinc/issues/4783)

but my OS is Windows11

**Comment by AenBleidd** [comment]
> > 如# 4783[的副本关闭](https://github.com/BOINC/boinc/issues/4783)
> 
> but my OS is Windows11

This happens on any OS and is already fixed.
Fixed version will be provided soon. Please stay tuned

**Comment by LezheGao** [comment]
> > > 如#4783[的副本关闭](https://github.com/BOINC/boinc/issues/4783)
> > 
> > 
> > 但我的操作系统是 Windows11
> 
> 这发生在任何操作系统上并且已经修复。 固定版本将很快提供。敬请期待

OK，Waiting...

**Comment by wangqr** [comment]
Came here from https://github.com/microsoft/vcpkg/issues/25437.
I can see the debug alert dialog using the [binary installer from 7.20.0 release](https://github.com/BOINC/boinc/releases/tag/client_release%2F7.20%2F7.20.0). But cannot see the alert if I build boinc from source f35e562 with MSVC, using win_build/boinc_vs2019.sln. How is the binary installer built?

**Comment by AenBleidd** [comment]
@wangqr, probably because you're using wxwidgets port with already set 'wxDEBUG_LEVEL=0' ?

**Comment by wangqr** [comment]
> @wangqr, probably because you're using wxwidgets port with already set 'wxDEBUG_LEVEL=0' ?

Yes. Once I add `git checkout dc6f878c0f49c3b0625d4f94217732fb3743b251` (one commit before microsoft/vcpkg#25240 was merged) after [`git pull`](https://github.com/BOINC/boinc/blob/client_release/7.20/7.20.0/win_build/vcpkg_3rdparty_dependencies_vs2019.vcxproj#L171), I can reproduce the issue.

The reason is that, wxWidgets is built with`wxDEBUG_LEVEL=1`, but boincmgr is built with `wxDEBUG_LEVEL=0`.

-----

> Here's how [`NDEBUG` disables asserts] works:
> * The entry point of a wxApp will call macro wxDISABLE_DEBUG_SUPPORT;([source](https://github.com/wxWidgets/wxWidgets/blob/v3.1.6/include/wx/app.h#L854))
> * wxDISABLE_DEBUG_SUPPORT has a call to wxDISABLE_ASSERTS_IN_RELEASE_BUILD;([source](https://github.com/wxWidgets/wxWidgets/blob/v3.1.6/include/wx/app.h#L40))
> * wxDISABLE_ASSERTS_IN_RELEASE_BUILD expands to wxDisableAsserts if NDEBUG is set;([source](https://github.com/wxWidgets/wxWidgets/blob/v3.1.6/include/wx/debug.h#L153))
> * wxDisableAsserts sets AssertHandler to NULL, disabling asserts.([source](https://github.com/wxWidgets/wxWidgets/blob/v3.1.6/include/wx/debug.h#L143))

So there are several cases:
* a) If both wxWidgets and boincmgr are built with `wxDEBUG_LEVEL=0`, debugging related code is completely removed. This is the current approach
* b) If both wxWidgets and boincmgr are built with `wxDEBUG_LEVEL=1` (or left as default), and boincmgr defines `NDEBUG`, debugging (assertions, debug logging) will be disabled upon wxApp init. This is, in my opinion, the prefered approach.
* c) If wxWidgets is built with `wxDEBUG_LEVEL=1`, but boincmgr is built with `wxDEBUG_LEVEL=0`. Then even if we define `NDEBUG`, debugging will not be disabled. This is because the header of wx assumes that there's no debugging functions present in wx, thus the `wxDisableAsserts` is [a dummy function](https://github.com/wxWidgets/wxWidgets/blob/v3.1.6/include/wx/debug.h#L133-L136), not correctly disabling debugging. This is how this issue happens.

So my opinion here is to left wxDEBUG_LEVEL as default (1) for both wxWidgets and boincmgr, and define `NDEBUG` in Release build of boincmgr. In my test this does fix this issue.

**Comment by AenBleidd** [comment]
@wangqr, if you check, 'NDEBUG' is already set for release builds of BOINC: https://github.com/BOINC/boinc/blob/e9d122480b8c54d6cd36e7b4e4e2513111f7aa9c/win_build/boincmgr_vs2019.vcxproj#L107
But debug asserts are still there and shown in runtime

**Comment by AenBleidd** [comment]
@wangqr, I assume, to keep asserts disabled, wxwidgets should be built with 'NDEBUG' also, but this flag is not supported by cmake build of wxwidgets.

**Comment by wangqr** [comment]
> But debug asserts are still there and shown in runtime

Yes, its because `wxDEBUG_LEVEL` is set to 0. If we remove that, i.e. changing that line to:
```
 <PreprocessorDefinitions>WIN32;NDEBUG;_WINDOWS;_UNICODE;UNICODE;wxUSE_GUI=1</PreprocessorDefinitions> 
```
Then this issue will be fixed

**Comment by AenBleidd** [comment]
@wangqr, so you want to say that these two flags contradict each other, and if I set 'wxDEBUG_LEVEL=0' in my app, then 'NDEBUG' flag will keep debug asserts enabled?
Either I don't understand smth or wxwidgets has broken logic.

**Comment by wangqr** [comment]
> @wangqr, so you want to say that these two flags contradict each other, and if I set 'wxDEBUG_LEVEL=0' in my app, then 'NDEBUG' flag will keep debug asserts enabled?

It is not 'wxDEBUG_LEVEL=0' and 'NDEBUG' contradict each other. It's 'wxDEBUG_LEVEL=1' when building wx, and 'wxDEBUG_LEVEL=0' in boincmgr, contradict each other.

* 'wxDEBUG_LEVEL=1' when building wx enables the debugging function in wx
* 'wxDEBUG_LEVEL=0' in boincmgr means: "We are linking a wx library without debugging functions, therefore we should not call any debugging related function, and call dummy functions instead."
* Normally when using wx in Release build, `NDEBUG` will set asserting handler to NULL, disabling assertions.
* With 'wxDEBUG_LEVEL=0' in boincmgr, we are not calling any debugging related function, including the one that disables assertions.

**Comment by AenBleidd** [comment]
@wangqr, then I suggest next: make this flag a port option. Then this will enable you to keep debug asserts dormant and I can completely remove them from my release builds.
What do you think?

**Comment by wangqr** [comment]
IIRC there is no "port option" in vcpkg. You may make it a "feature" that can be toggled on or off. This will work for my use case.

**Comment by AenBleidd** [comment]
@wangqr, sorry, my bad. Saying 'option' I meant 'feature'. Thank you for correcting me.
